MdeModulePkg/Core/Dxe/Misc/MemoryProtectionSupport.c

```diff
@@ -265,74 +265,290 @@ SetUefiImageMemoryAttributes (
+  }
+
+  ImageRecordToInsertLink->BackLink              = ImageRecordLink->BackLink;
+  ImageRecordToInsertLink->ForwardLink           = ImageRecordLink;
+  ImageRecordToInsertLink->BackLink->ForwardLink = ImageRecordToInsertLink;
+  ImageRecordToInsertLink->ForwardLink->BackLink = ImageRecordToInsertLink;
+  return EFI_SUCCESS;
+}
+
+/**
+  Inserts the input CodeSectionToInsertLink into CodeSectionList based on the
+  IMAGE_PROPERTIES_RECORD_CODE_SECTION.CodeSegmentBase field
+
+  @param[in] CodeSectionList           Pointer to the head of the IMAGE_PROPERTIES_RECORD_CODE_SECTION list
+  @param[in] CodeSectionToInsertLink   Pointer to the list entry of the IMAGE_PROPERTIES_RECORD_CODE_SECTION to insert
+
+  @retval EFI_SUCCESS             IMAGE_PROPERTIES_RECORD_CODE_SECTION inserted into the list
+  @retval EFI_INVALID_PARAMETER   CodeSectionList or CodeSectionToInsertLink were NULL
+**/
+STATIC
+EFI_STATUS
+OrderedInsertCodeSectionListEntry (
+  IN LIST_ENTRY  *CodeSectionList,
+  IN LIST_ENTRY  *CodeSectionToInsertLink
+  )
+{
+  IMAGE_PROPERTIES_RECORD_CODE_SECTION  *CurrentCodeSection;
+  IMAGE_PROPERTIES_RECORD_CODE_SECTION  *CodeSectionToInsert;
+  LIST_ENTRY                            *CodeSectionLink;
+  LIST_ENTRY                            *CodeSectionEndLink;
+  EFI_PHYSICAL_ADDRESS                  CodeSectionToInsertBase;
 
-      NextSpecialRegionEntryLink = NextSpecialRegionEntryLink->ForwardLink;
+  if ((CodeSectionList == NULL) || (CodeSectionToInsertLink == NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  CodeSectionToInsert = CR (
+                          CodeSectionToInsertLink,
+                          IMAGE_PROPERTIES_RECORD_CODE_SECTION,
+                          Link,
+                          IMAGE_PROPERTIES_RECORD_CODE_SECTION_SIGNATURE
+                          );
+
+  CodeSectionToInsertBase = CodeSectionToInsert->CodeSegmentBase;
+
+  CodeSectionLink    = CodeSectionList->ForwardLink;
+  CodeSectionEndLink = CodeSectionList;
+  while (CodeSectionLink != CodeSectionEndLink) {
+    CurrentCodeSection = CR (
+                           CodeSectionLink,
+                           IMAGE_PROPERTIES_RECORD_CODE_SECTION,
+                           Link,
+                           IMAGE_PROPERTIES_RECORD_CODE_SECTION_SIGNATURE
+                           );
+    if (CodeSectionToInsertBase < CurrentCodeSection->CodeSegmentBase) {
+      break;
     }
 
-    SpecialRegionEntryLink     = SpecialRegionEntryLink->ForwardLink;
-    NextSpecialRegionEntryLink = SpecialRegionEntryLink->ForwardLink;
+    CodeSectionLink = CodeSectionLink->ForwardLink;
   }
+
+  CodeSectionToInsertLink->BackLink              = CodeSectionLink->BackLink;
+  CodeSectionToInsertLink->ForwardLink           = CodeSectionLink;
+  CodeSectionToInsertLink->BackLink->ForwardLink = CodeSectionToInsertLink;
+  CodeSectionToInsertLink->ForwardLink->BackLink = CodeSectionToInsertLink;
+  return EFI_SUCCESS;

```

Comment:
This appears to be the same function three times for the different types. Does it make sense to make a generic function to call? It gets a little bit uglier with having to pass the type in, but if the types are similar enough in the parts that are getting rearranged a simple abstract struct could be used?

If it's too difficult or I'm missing there are differences in these ordered insert functions, then we can disregard.

Reply:
Hmmmm.... I think maybe if the function only compares UINT64 values, then I can pass in the offset of the value being compared, the offset of that expected signature, and the expected signature itself to create an insert function.

Reply:
Yeah it works - I'll consolidate them

Reply:
Please review: https://github.com/microsoft/mu_basecore/pull/148/files